Implement result() function in MPI Work classes

agolynski opened the pull request:
Implement result() for MPI which wasn't previously supported.

Some user rely on output args, however in future usecases (e.g. DDP comm hook) we need to return the result explicitly.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57168 Implement result() function in MPI Work classes**

Differential Revision: [D28129125](https://our.internmc.facebook.com/intern/diff/D28129125)

agolynski committed:
Implement return() function in MPI Work

[ghstack-poisoned]

agolynski committed:
Update on "Implement return() function in MPI Work"




[ghstack-poisoned]

agolynski committed:
Update on "Implement return() function in MPI Work"




[ghstack-poisoned]

agolynski commented:
cc @mrshenli

agolynski committed:
Update on "Implement return() function in MPI Work"




[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/test/ProcessGroupMPITest.cpp:
```diff
@@ -196,7 +207,7 @@ void testGather(int iter = 10000) {
       }
     } else {
-      allOutputTensors[i] = std::vector<std::vector<at::Tensor>>(1);
+      allOutputTensors[i] = std::vector<std::vector<at::Tensor>>(0);
```
this means initializing all tensor values to 0, right?

mrshenli reviewed (commented):
Could you please add a PR summary to describe what is changed by this PR and why?

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.hpp:
```diff
@@ -42,7 +42,11 @@ struct WorkEntry {
+
+  // We use `dst` for returning resulting vectors in WorkMPI::result, so we need to
+  // keep it alive.
+  std::shared_ptr<std::vector<at::Tensor>> dst;
```
could you please elaborate why the original `std::vector<at::Tensor>` was not sufficient? Is it because you wanna avoid making a copy of the vector when retrieving the outputs?

agolynski replied:
this issue was for vectors of tensors.

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.hpp:
```diff
@@ -77,24 +81,31 @@ class ProcessGroupMPI : public ProcessGroup {
               profilingTitle,
-              inputTensors) {}
+              inputTensors),
+          outputs_(outputs) {}
```
curious, why `const std::shared_ptr` + copy instead of `std::shared_ptr` + move?

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.hpp:
```diff
@@ -111,10 +122,12 @@ class ProcessGroupMPI : public ProcessGroup {
     void populateException();
 
-    at::Tensor tensor_;
+    const std::vector<at::Tensor>* outputs_;
```
curious, why we previously use a single tensor, but now need a vector of tensors?

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.cpp:
```diff
@@ -94,13 +94,21 @@ void checkSameSizeAndType(
 
+std::vector<at::Tensor> ProcessGroupMPI::WorkMPI::result() {
+  if (outputs_) {
+    return *outputs_;
```
I don't understand this. Since we are making a copy of the vector anyway, why do we need shared pointer?

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.cpp:
```diff
@@ -815,8 +832,8 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupMPI::send(
   return c10::make_intrusive<AsyncWork>(
-      tensor,
       request,
+      nullptr,
```
hmm, any reason that output tensors is of type `std::vector<at::Tensor>*` while inputTensors is of type `optional<std::vector<at::Tensor>>`?

mrshenli replied:
> we don't need to copy outputs here as user of send/recv manages lifetime of tensors

does this mean inputs' lifetime is managed differently, so that we need to copy inputs?

agolynski replied:
I think we do not specify this in our API directly.

I've updated `dst` management: since none of the ops directly modify it, we can make it const and hence WorkMPI and WorkEntry can each have their own copy of outputTensors_.

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.hpp:
```diff
@@ -77,24 +81,31 @@ class ProcessGroupMPI : public ProcessGroup {
   class WorkMPI : public ProcessGroup::Work {
    public:
     WorkMPI(
+        const std::shared_ptr<std::vector<at::Tensor>> outputs,
```
Is it intentional to have outputs/inputTensors instead of ouputTensors/inputTensors?

agolynski committed:
Update on "Implement return() function in MPI Work"




[ghstack-poisoned]

agolynski commented on torch/lib/c10d/test/ProcessGroupMPITest.cpp:
```diff
@@ -36,13 +41,14 @@ void testAllreduce(int iter = 1000) {
 
   std::vector<c10::intrusive_ptr<::c10d::ProcessGroup::Work>> works;
   for (auto& tensors : allTensors) {
-    // Kick off work
```
I thought it's misleading since it = puts work in queue and doesn't kick it off. the line is 
work = pg->allreduce(tensors);
is self-explanatory.

Changed comments here and in the rest of module

agolynski committed:
Update on "Implement return() function in MPI Work"




[ghstack-poisoned]

agolynski committed:
Update on "Implement return() function in MPI Work"




[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/test/ProcessGroupMPITest.cpp:
```diff
@@ -14,14 +14,19 @@
     c10::intrusive_ptr<::c10d::ProcessGroupMPI> pg,
-    std::vector<c10::intrusive_ptr<c10d::ProcessGroup::Work>> works) {
+    std::vector<c10::intrusive_ptr<c10d::ProcessGroup::Work>> works,
+    std::vector<std::vector<at::Tensor>>* outputTensors) {
```
is there any downside of always returning a vector? Since this is only for test, I guess perf is not a problem?

agolynski replied:
> > Could you please add a PR summary to describe what is changed by this PR and why?
> 
> Just wanna mention that this hasn't been addressed yet.

sorry , haven't noticed this. updated the description.

mrshenli commented on torch/lib/c10d/test/ProcessGroupMPITest.cpp:
```diff
@@ -36,13 +41,14 @@ void testAllreduce(int iter = 1000) {
 
   std::vector<c10::intrusive_ptr<::c10d::ProcessGroup::Work>> works;
   for (auto& tensors : allTensors) {
```
The old code needs `allTensor` because it needs to verify outputs using `allTensor`. Now the new code get output directly from `waitWork()`, why do we still need `allTensors`? Will it work if we just construct tensors and pass them to `pg->allreduce` in this loop? Or is it because the application code should keep the tensors alive and so that temporary tensors won't work?

agolynski committed:
Update on "Implement return() function in MPI Work"




[ghstack-poisoned]

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.cpp:
```diff
@@ -94,13 +94,17 @@ void checkSameSizeAndType(
+  return outputTensors_;
+}
+
 ProcessGroupMPI::AsyncWork::AsyncWork(
```
Could you please add a paragraph to describe how tensor lifetime and ownership is managed among 1) user code 2) AsyncWork 3) WorkMPI 4) WorkEntry? Besides that, it will be super helpful to use send/recv and allreduce as two examples.

agolynski replied:
I don't understand: tensors lifetime and ownership is handled by tensor library ref counting. we do no introduce anything new here?

for vectors: is it still relevant? right now each of AsyncWork, WorkMPI, WorkEntry keeps its own copy of user provided outputs.

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.cpp:
```diff
@@ -844,8 +858,8 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupMPI::recv(
   }
 
   return c10::make_intrusive<AsyncWork>(
```
(This is prior to this PR and can be addressed in followup PRs) Would I be correct if I assume `AsyncWork` is only used in send/recv and allToAll, while all other API return `WorkMPI`? Why do we need two different work types?

agolynski replied:
AFAIK `AsyncWork` is only used in send/recv, impls of Alltoall are using sync MPI ops and WorkMPI.

nonblocking MPI calls are used with Async work, and blocking with WorkMPI. I am not sure why we made a decision of using nonblocking versions of send/recv.

There is some guidance on how to use MPI send/recv:
https://www.mcs.anl.gov/research/projects/mpi/sendmode.html

mrshenli replied:
Do we plan to keep `WorkMPI` and `AsyncWork` separate or does it make sense to merge? IIUC, we plan to use thread-pools to mark async send/recv as ready. If that's the case, it will bring the use cases of AsyncWork closer to WorkMPI?

agolynski replied:
correct. i think it'd make sense, in this case we won't have benefits of MPI_ISend/MPI_IRecv and should probably switch to their sync counterparts.

mrshenli commented on torch/lib/c10d/ProcessGroupMPI.hpp:
```diff
@@ -111,10 +120,12 @@ class ProcessGroupMPI : public ProcessGroup {
     void populateException();
 
-    at::Tensor tensor_;
+    const std::vector<at::Tensor>* outputTensors_;
```
Looks like both `AsyncWork` and `WorkMPI` are user-facing Work implementations. Why they own `outputTensors` in different way?

mrshenli commented on torch/lib/c10d/test/ProcessGroupMPITest.cpp:
```diff
@@ -36,13 +41,14 @@ void testAllreduce(int iter = 1000) {
-  waitWork(pg, works);
+  std::vector<std::vector<at::Tensor>> outputTensors;
+
+  waitWork(pg, works, &outputTensors);
```
Looks OK to me. Just curious, what's the benefit of adding the `need_output` flag, instead of always returning the output tensors? Is the intention to avoid populating the result vector and save some overhead?

agolynski replied:
no, i was thinking that we might not want to define work->result() in case no result is supposed to be present. however, we are returning empty list of tensors in this case.

mrshenli reviewed (commented):
> Could you please add a PR summary to describe what is changed by this PR and why?

Just wanna mention that this hasn't been addressed yet.

agolynski committed:
Update on "Implement return() function in MPI Work"

Implement return() for MPI which wasn't previously supported.

Some user rely on output args, however in future usecases (e.g. DDP comm hook) we need to return the result explicitly.




[ghstack-poisoned]

agolynski committed:
Update on "Implement return() function in MPI Work"

Implement return() for MPI which wasn't previously supported.

Some user rely on output args, however in future usecases (e.g. DDP comm hook) we need to return the result explicitly.




[ghstack-poisoned]

agolynski committed:
Update on "Implement return() function in MPI Work"

Implement return() for MPI which wasn't previously supported.

Some user rely on output args, however in future usecases (e.g. DDP comm hook) we need to return the result explicitly.




[ghstack-poisoned]

mrshenli reviewed (commented):
> Implement return() for MPI which wasn't previously supported.

Do you mean `result()`?

mrshenli reviewed (commented):
build failures are real (merge conflicts). Please rebase

agolynski committed:
Update on "Implement return() function in MPI Work"

Implement result() for MPI which wasn't previously supported.

Some user rely on output args, however in future usecases (e.g. DDP comm hook) we need to return the result explicitly.




[ghstack-poisoned]

agolynski commented:
> > Could you please add a PR summary to describe what is changed by this PR and why?
> 
> Just wanna mention that this hasn't been addressed yet.

done

agolynski committed:
Update on "Implement result() function in MPI Work classes"

Implement result() for MPI which wasn't previously supported.

Some user rely on output args, however in future usecases (e.g. DDP comm hook) we need to return the result explicitly.


Differential Revision: [D28129125](https://our.internmc.facebook.com/intern/diff/D28129125)

[ghstack-poisoned]

agolynski committed:
Update on "Implement result() function in MPI Work classes"

Implement result() for MPI which wasn't previously supported.

Some user rely on output args, however in future usecases (e.g. DDP comm hook) we need to return the result explicitly.


Differential Revision: [D28129125](https://our.internmc.facebook.com/intern/diff/D28129125)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM!
